### PR TITLE
Feature/smol repay

### PR DESCRIPTION
**Pull request by kjaredic**

### CODE REVIEW

File: contracts/actions/mcd/McdBoostComposite.sol [hunk 1]
```diff
@@ -0,0 +1,248 @@
+// SPDX-License-Identifier: SEE LICENSE IN LICENSE
```
Comment (majkic99) [inline contracts/actions/mcd/McdBoostComposite.sol:1]: MIT

Comment (majkic99) [inline contracts/actions/mcd/McdBoostComposite.sol:1]: MIT

File: contracts/actions/mcd/McdBoostComposite.sol [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+ActionBase, DFSSell, McdHelper, IFlashLoanRecipient,
+ReentrancyGuard, MainnetBalancerV2Addresses {
+    using TokenUtils for address;
+
+    address internal immutable ACTION_ADDR = address(this);
+
+    /// @param _vaultId Id of the vault
+    /// @param _amount Amount of dai to be payed back
+    /// @param _from Where the Dai is pulled from
+    /// @param _mcdManager The manager address we are using
```
Comment (majkic99) [inline contracts/actions/mcd/McdBoostComposite.sol:30]: params natspec is off

File: contracts/actions/mcd/McdRepayComposite.sol [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+    /// @notice Returns the type of action we are implementing
+    function actionType()
+    public
+    pure
+    virtual override (ActionBase, DFSSell)
+    returns (uint8) {
+        return uint8(ActionType.CUSTOM_ACTION);
+    }
+
+    function _parseCompositParams(bytes memory _calldata) internal pure returns (RepayParams memory params) {
```
Comment (majkic99) [inline contracts/actions/mcd/McdRepayComposite.sol:260]: add E to composit

File: contracts/actions/mcd/McdRepayComposite.sol [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+                    _repayParams.joinAddr,
+                    paybackAmount,
+                    _repayParams.repayAmount
+                );
+            }
+            collateralAssetBalanceDelta = joinToken.getBalance(address(this)) - collateralAssetBalanceDelta;
+            require(collateralAssetBalanceDelta == _repayParams.repayAmount);
+        }
+
+        logger.logRecipeEvent("McdRepayComposite");
```
Comment (majkic99) [inline contracts/actions/mcd/McdRepayComposite.sol:188]: unsure if these two should be split, one in action direct one in recipe execution

File: contracts/actions/mcd/McdRepayComposite.sol [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+        uint256 flPaybackAmount = _amounts[0] + _feeAmounts[0];
+        _tokens[0].withdrawTokens(address(VAULT_ADDR), flPaybackAmount);
+    }
+
+    function _repay(address _proxy, RepayParams memory _repayParams) internal {
+        uint256 collateral = getAllColl(IManager(_repayParams.mcdManager), _repayParams.joinAddr, _repayParams.vaultId);
+        _repayParams.repayAmount = _repayParams.repayAmount > collateral ? collateral : _repayParams.repayAmount;
+
+        // Sell flashloaned collateral asset for debt asset
+        assert(_repayParams.exchangeData.srcAmount == _repayParams.repayAmount); // CONSIDER
```
Comment (majkic99) [inline contracts/actions/mcd/McdRepayComposite.sol:144]: makes sense, but sending different srcAmount and repayAmount should never happen, what about setting srcAmount to repayAmount?

File: test/mcd/mcd-tests.js [hunk 6]
```diff
@@ -930,6 +939,330 @@ const mcdClaimTest = async () => {
...
+                expect(await balanceOf(tokenData.address, boostComposite.address)).to.be.eq(0);
+                expect(await balanceOf(DAI_ADDR, boostComposite.address)).to.be.eq(0);
+                expect(await balanceOf(tokenData.address, proxy.address)).to.be.eq(0);
+                expect(await balanceOf(DAI_ADDR, proxy.address)).to.be.eq(0);
+                expect(feesAfter.sub(feesBefore)).to.be.eq(feeAmount);
+            });
+        }
+    });
+};
+
```
Comment (majkic99) [inline test/mcd/mcd-tests.js:1265]: would create a test where a recipe includes composite actions just to be on the safe side

File: contracts/actions/mcd/McdBoostComposite.sol [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+    function executeActionDirect(bytes memory _callData) public payable virtual override (ActionBase, DFSSell) {
+        BoostParams memory boostParams = _parseCompositParams(_callData);
+        _flBalancer(boostParams);
+    }
+
+    /// @notice Gets a FL from Balancer and returns back the execution to the action address
+    function _flBalancer(BoostParams memory _boostParams) internal {
+        address[] memory tokens = new address[](1);
+        uint256[] memory amounts = new uint256[](1);
+        tokens[0] = _boostParams.exchangeData.srcAddr;
```
Comment (majkic99) [inline contracts/actions/mcd/McdBoostComposite.sol:109]: can be hardcoded to DAI? or have a check if .srcAddr is not DAI to revert on start to not spend more gas than needed later on

File: contracts/actions/mcd/McdRepayComposite.sol (hunk 5 above)
Comment (kjaredic) [reply contracts/actions/mcd/McdRepayComposite.sol]: removed unneeded repayAmount param, srcAmount used instead

File: contracts/actions/mcd/McdBoostComposite.sol (hunk 7 above)
Comment (kjaredic) [reply contracts/actions/mcd/McdBoostComposite.sol:103]: makes sense

**Review by majkic99 — COMMENTED** [review]
lgtm

**Comment by yippee-ki-yay** [comment]
Is part of another pr

**Pull request closed without merging**